mplc/multi_partner_learning.py

```diff
@@ -654,3 +655,151 @@ def init_multi_partner_learning_from_scenario(scenario, is_save_data=True):
+        self.history_theta_ = [[None for _ in self.partners_list] for _ in range(self.epoch_count)]
+        self.theta = self.init_flip_proba()
+        self.theta_ = None
+        # self.labels_map = LabelEncoder().fit_transform([str(y) for y in scenario.dataset.y_train])
+
+    def init_flip_proba(self):
+        identity = np.identity(self.K)
+        return identity * (1 - self.epsilon) + (1 - identity) * (self.epsilon / (self.K - 1))
+
+    def fit(self):
+        start = timer()
+        logger.info(
+            f"## Training and evaluating model on partners with ids: {['#' + str(p.id) for p in self.partners_list]}")
+
+        logger.info("(LFlip) Very first minibatch, init new models for each partner")
+        partners_models = self.init_with_models()
+
+        while self.epoch_index < self.epoch_count:
+            self.split_in_minibatches()
+            self.minibatch_index = 0
+            while self.minibatch_index < self.minibatch_count:
+                logger.info(f"(LFLip) Minibatch n°{self.minibatch_index} of epoch n°{self.epoch_index}, "
+                            f"init aggregated model for each partner with models from previous round")
+
+                # Evaluate and store accuracy of mini-batch start model
+                model_to_evaluate = partners_models[0]
+                model_evaluation_val_data = self.evaluate_model(model_to_evaluate, self.val_data)
+                self.score_matrix_collective_models[self.epoch_index,
+                                                    self.minibatch_index] = model_evaluation_val_data[1]
+
+                for partner_index, partner in enumerate(self.partners_list):
+                    # Reference the partner's model
+                    partner_model = partners_models[partner_index]
+                    x_batch = self.minibatched_x_train[partner_index][self.minibatch_index]
+                    y_batch = self.minibatched_y_train[partner_index][self.minibatch_index]
+
+                    predictions = partner_model.predict(x_batch)
+                    self.theta_ = predictions  # Initialize the theta_
+
+                    for idx, y in enumerate(y_batch):
+                        self.theta_[idx, :] *= self.theta[:, np.argmax(y)]
+                        self.theta_ = normalize(self.theta_, axis=1, norm='l1')
+                    self.history_theta_[self.epoch_index][partner_index] = self.theta
+
+                    self.theta = self.theta_.T.dot(y_batch)
+                    self.theta = normalize(self.theta, axis=1, norm='l1')
+
+                    self.history_theta[self.epoch_index][partner_index] = self.theta
+
+                    self.theta_ = predictions
+                    for idx, y in enumerate(y_batch):
+                        self.theta_[idx, :] *= self.theta[:, np.argmax(y)]
+                        self.theta_ = normalize(self.theta_, axis=1, norm='l1')
```

Comment:
Same thing, line 732 can be put out of the for loop
```suggestion
                    for idx, y in enumerate(y_batch):
                        self.theta_[idx, :] *= self.theta[:, np.argmax(y)]
                    self.theta_ = normalize(self.theta_, axis=1, norm='l1')
```